File: go/suppressions.go
package sentroy

import (
	"net/http"
	"net/url"
	"strconv"
)

// SuppressionsService manages the suppression list. Suppressed recipients
// are skipped at send time.
type SuppressionsService struct {
	h *httpClient
}

func newSuppressionsService(h *httpClient) *SuppressionsService {
	return &SuppressionsService{h: h}
}

// List returns suppressions across the company (or a single domain). Every
// entry is one address that will not receive mail until removed.
func (s *SuppressionsService) List(params *SuppressionListParams) ([]Suppression, error) {
	q := map[string]string{}
	if params != nil {
		if params.Page > 0 {
			q["page"] = strconv.Itoa(params.Page)
		}
		if params.Limit > 0 {
			q["limit"] = strconv.Itoa(params.Limit)
		}
		if params.DomainID != "" {
			q["domainId"] = params.DomainID
		}
		if params.Reason != "" {
			q["reason"] = params.Reason
		}
	}
	return doRequest[[]Suppression](s.h, http.MethodGet, "/suppressions", q, nil)
}

// Add manually suppresses an address (e.g. honoring an off-platform
// opt-out). Bounces and complaints are added automatically by the mail
// server.
func (s *SuppressionsService) Add(params AddSuppressionParams) (*Suppression, error) {
	sup, err := doRequest[Suppression](s.h, http.MethodPost, "/suppressions", nil, params)
	if err != nil {
		return nil, err
	}
	return &sup, nil
}

// Remove deletes a suppression — the address will be eligible to receive
// mail again.
func (s *SuppressionsService) Remove(id string) error {
	_, err := doRequest[map[string]interface{}](s.h, http.MethodDelete,
		"/suppressions/"+url.PathEscape(id), nil, nil)
	return err
}

File: php/src/Suppressions.php
<?php

namespace Sentroy\ClientSdk;

class Suppressions
{
    /** @var HttpClient */
    private $http;

    public function __construct(HttpClient $http)
    {
        $this->http = $http;
    }

    /**
     * List suppressions across the company (or a single domain).
     * Suppressed recipients are skipped at send time — every entry here
     * is one address that will not receive mail until removed.
     *
     * @param array $params {
     *     @type int    $page     Optional
     *     @type int    $limit    Optional
     *     @type string $domainId Optional
     *     @type string $reason   Optional
     * }
     * @return array
     */
    public function getAll(array $params = array())
    {
        $query = array();
        if (isset($params['page'])) {
            $query['page'] = $params['page'];
        }
        if (isset($params['limit'])) {
            $query['limit'] = $params['limit'];
        }
        if (isset($params['domainId'])) {
            $query['domainId'] = $params['domainId'];
        }
        if (isset($params['reason'])) {
            $query['reason'] = $params['reason'];
        }
        return $this->http->get('/suppressions', $query);
    }

    /**
     * Manually suppress an address (e.g. honoring an off-platform
     * opt-out). Bounces and complaints are added automatically by the
     * mail server.
     *
     * @param array $params {
     *     @type string $email    Required
     *     @type string $reason   Optional
     *     @type string $domainId Required
     * }
     * @return array
     */
    public function add(array $params)
    {
        return $this->http->post('/suppressions', $params);
    }

    /**
     * Remove a suppression — the address will be eligible to receive
     * mail again.
     *
     * @param string $id
     * @return void
     */
    public function remove($id)
    {
        $this->http->delete('/suppressions/' . rawurlencode($id));
    }
}
